### PR TITLE
Fix a problem in the time zone info generation

### PR DESCRIPTION
**Pull request by ehoogerbeets**
- The java zone info compiler (zic) was not handling time zones
properly that had 3 levels of name, like "America/Indiana/Marengo"
- This caused all time zones in Indiana to overwrite the same
file, which meant that some time zones were just missing
- The zonetab.json was not sorted, making it difficult to see
problems in git diff output
- A few time zones had the wrong abbreviation because of an edge
case in the zoneinfo definitions from IANA

### Checklist

* [x] At least one test case is included for this feature or bug fix.
* [x] `ReleaseNotes` has added or is not needed.
* [x] This PR has passed all of the test cases on `Nodejs` and `Chrome Browser`.
* [ ] This PR has passed all of the test cases on `QT/QML`.
* [ ] This is an API breaking change.
* [ ] Requires a major version change.

### Issue Resolved / Feature Added
[//]: # (Describe the issue resolved or feature added by this pull request)

### Resolution
[//]: # (Does the code work as intended?)
[//]: # (What is the impact of this change and *why* was it made?)

### Links
[//]: # (Related issues, references)

### CODE REVIEW

**Comment by gouniLee** [comment]
Did you just regenerating the `tzdatac` which is currently iLib based on? It looks like many of zoninfo data has been up updated not only just Indiana.

File: js/test/calendar/testpersiandateastro.js [hunk 1]
```diff
@@ -635,7 +635,7 @@ module.exports.testpersiandateastro = {
 
         test.ok(pd !== null);
 
-        test.equal(pd.getJulianDay(), 2458199.238194444);
+        test.equal(pd.getJulianDay(), 2458199.1965277777);
```
Comment (gouniLee) [inline js/test/calendar/testpersiandateastro.js]: This zoneinfo data changing affects these test cases too?

Comment (ehoogerbeets) [reply js/test/calendar/testpersiandateastro.js]: The given persian date is 3/21/2018 in Gregorian. The dates of start and end of DST in the zone info file are in gregorian. The dates of the start and end of DST were calculated wrong before for Persian and were corrected from 3/22 to 3/21. That means that previously, the given date was before the start of DST and with this change, it is after the start of DST. That is the difference in Julian day calculation -- 1 hour. (2458199.238194444 - 2458199.1965277777) * 24 = 1

**Comment by ehoogerbeets** [comment]
I did regenerate the tzdata based on 2019c, which oddly is still the latest right now. Many of the files were overwriting `zoneinfo/America/Indiana.json` for example because of the miscalculation of the file path when the zone name had 3 levels instead of two. (ie. `America/Indiana/Indianapolis` vs. `America/Los_Angeles`). Now those files are separate, and `Indiana.json` is now.

Also, fixed a bug for an edge case where the zone definition had just plain "%s" for the abbreviation, and the full abbreviation was given in the rule definition instead. Most zones have something like "P%sT" where the %s is replaced by "S" or "D" depending on DST. That changed a few zones, like Namibia from CAT to WAT instead.

For other zones, like Casablanca, the start and end of DST is based on a different calendar. Rather than force implementors to be able to calculate things in that other calendar, the start and end of DST is recalculated in Gregorian for each year. Now that it is a year later, those dates for 2020 have changed. ie. a different rule applies.

Currently, I am working on rewriting the zic to fix these things. The benefits include:

1. Ability to support historical and future dates properly, so that we can support Casablanca and other non-gregorian calendars correctly
2. All JS tool... get rid of one more piece of Java in the build.
3. Simplify the generation of zone info because it will all be in 1 tool, instead of first running a java tool followed by a js tool.

**Comment by gouniLee** [comment]
> For other zones, like Casablanca, the start and end of DST is based on a different calendar. Rather than force implementors to be able to calculate things in that other calendar, the start and end of DST is recalculated in Gregorian for each year. Now that it is a year later, those dates for 2020 have changed. ie. a different rule applies.

As far as I know, we have an issue regarding in Casablanca timezone (Including Africa/Windhoek and Dublin) The issue was regarding negative dst (The Issue reported to chromium and they fixed it)  I don't get your saying about `the start and end of DST is based on a different calendar`, even in the same zone, it uses different calendar?

**Comment by gouniLee** [comment]
> I did regenerate the tzdata based on 2019c, which oddly is still the latest right now. Many of the files were overwriting `zoneinfo/America/Indiana.json` for example because of the miscalculation of the file path when the zone name had 3 levels instead of two. (ie. `America/Indiana/Indianapolis` vs. `America/Los_Angeles`). Now those files are separate, and `Indiana.json` is now.
> 
> Also, fixed a bug for an edge case where the zone definition had just plain "%s" for the abbreviation, and the full abbreviation was given in the rule definition instead. Most zones have something like "P%sT" where the %s is replaced by "S" or "D" depending on DST. That changed a few zones, like Namibia from CAT to WAT instead.
> 
> For other zones, like Casablanca, the start and end of DST is based on a different calendar. Rather than force implementors to be able to calculate things in that other calendar, the start and end of DST is recalculated in Gregorian for each year. Now that it is a year later, those dates for 2020 have changed. ie. a different rule applies.
> 
> Currently, I am working on rewriting the zic to fix these things. The benefits include:
> 
> 1. Ability to support historical and future dates properly, so that we can support Casablanca and other non-gregorian calendars correctly
> 2. All JS tool... get rid of one more piece of Java in the build.
> 3. Simplify the generation of zone info because it will all be in 1 tool, instead of first running a java tool followed by a js tool.

**Comment by gouniLee** [comment]
oooooppps Sorry, ;( accidentally clicked the close button. so I reopen it ^^;;;;;;;;;;;;;;

**Comment by ehoogerbeets** [comment]
In Morocco, they decide on the start and end of DST based on the Arabic calendar. It is the same date in the Arabic calendar each year, but as you know, that calendar is lunisolar, so the same date in the Arabic calendar is a different Gregorian date each year. Rather than making implementors write Arabic calendar code just to figure out DST, they pre-calculated the Gregorian dates when DST starts and ends and make a special rule for each year. Here is a small sampling of the Morocco rules:

```
Rule    Morocco 2020    only    -       Apr     19       3:00   -1:00   -
Rule    Morocco 2020    only    -       May     24       2:00   0       -
Rule    Morocco 2021    only    -       Apr     11       3:00   -1:00   -
Rule    Morocco 2021    only    -       May     16       2:00   0       -
Rule    Morocco 2022    only    -       Mar     27       3:00   -1:00   -
Rule    Morocco 2022    only    -       May      8       2:00   0       -
Rule    Morocco 2023    only    -       Mar     19       3:00   -1:00   -
Rule    Morocco 2023    only    -       Apr     23       2:00   0       -
Rule    Morocco 2024    only    -       Mar     10       3:00   -1:00   -
Rule    Morocco 2024    only    -       Apr     14       2:00   0       -
Rule    Morocco 2025    only    -       Feb     23       3:00   -1:00   -
Rule    Morocco 2025    only    -       Apr      6       2:00   0       -
Rule    Morocco 2026    only    -       Feb     15       3:00   -1:00   -
Rule    Morocco 2026    only    -       Mar     22       2:00   0       -
Rule    Morocco 2027    only    -       Feb      7       3:00   -1:00   -
Rule    Morocco 2027    only    -       Mar     14       2:00   0       -
Rule    Morocco 2028    only    -       Jan     23       3:00   -1:00   -
Rule    Morocco 2028    only    -       Feb     27       2:00   0       -
Rule    Morocco 2029    only    -       Jan     14       3:00   -1:00   -
Rule    Morocco 2029    only    -       Feb     18       2:00   0       -
Rule    Morocco 2029    only    -       Dec     30       3:00   -1:00   -
Rule    Morocco 2030    only    -       Feb     10       2:00   0       -
Rule    Morocco 2030    only    -       Dec     22       3:00   -1:00   -
```

**Comment by gouniLee** [comment]
> In Morocco, they decide on the start and end of DST based on the Arabic calendar. It is the same date in the Arabic calendar each year, but as you know, that calendar is lunisolar, so the same date in the Arabic calendar is a different Gregorian date each year. Rather than making implementors write Arabic calendar code just to figure out DST, they pre-calculated the Gregorian dates when DST starts and ends and make a special rule for each year. Here is a small sampling of the Morocco rules:
> 
> ```
> Rule    Morocco 2020    only    -       Apr     19       3:00   -1:00   -
> Rule    Morocco 2020    only    -       May     24       2:00   0       -
> Rule    Morocco 2021    only    -       Apr     11       3:00   -1:00   -
> Rule    Morocco 2021    only    -       May     16       2:00   0       -
> Rule    Morocco 2022    only    -       Mar     27       3:00   -1:00   -
> Rule    Morocco 2022    only    -       May      8       2:00   0       -
> Rule    Morocco 2023    only    -       Mar     19       3:00   -1:00   -
> Rule    Morocco 2023    only    -       Apr     23       2:00   0       -
> Rule    Morocco 2024    only    -       Mar     10       3:00   -1:00   -
> Rule    Morocco 2024    only    -       Apr     14       2:00   0       -
> Rule    Morocco 2025    only    -       Feb     23       3:00   -1:00   -
> Rule    Morocco 2025    only    -       Apr      6       2:00   0       -
> Rule    Morocco 2026    only    -       Feb     15       3:00   -1:00   -
> Rule    Morocco 2026    only    -       Mar     22       2:00   0       -
> Rule    Morocco 2027    only    -       Feb      7       3:00   -1:00   -
> Rule    Morocco 2027    only    -       Mar     14       2:00   0       -
> Rule    Morocco 2028    only    -       Jan     23       3:00   -1:00   -
> Rule    Morocco 2028    only    -       Feb     27       2:00   0       -
> Rule    Morocco 2029    only    -       Jan     14       3:00   -1:00   -
> Rule    Morocco 2029    only    -       Feb     18       2:00   0       -
> Rule    Morocco 2029    only    -       Dec     30       3:00   -1:00   -
> Rule    Morocco 2030    only    -       Feb     10       2:00   0       -
> Rule    Morocco 2030    only    -       Dec     22       3:00   -1:00   -
> ```

oh.. I see, they calculate dst based on Arabic calendar. now I get that saying from IANA release note
https://data.iana.org/time-zones/tzdb-2019a/NEWS
> Morocco is scheduled to observe three DST transitions in some Gregorian years (e.g., 2033) due to the mismatch between the Gregorian and Islamic calendars. Thank you!

and Good to know that you're writing code with js :)

**Pull request merged**